File: leetcode/algorithms/stack/daily-temperatures.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Daily Temperatures"
*/

class Record {
    public int dayIndex;
    public int temperature;
    public Record(int _index,int _temp){
        dayIndex = _index;
        temperature = _temp;
    }
}


class Solution {
    public int[] dailyTemperatures(int[] T){
        int[] ansArr = new int[T.length];
        
        Stack<Record> tempStack = new Stack<Record>();
        
        for(int i=T.length-1;i>=0;--i){
            while(!tempStack.empty() && tempStack.peek().temperature<=T[i]){
                tempStack.pop();
            }
            if(!tempStack.empty()){
                ansArr[i] = tempStack.peek().dayIndex - i;
            } else {
                ansArr[i] = 0;
            }
            
            tempStack.push(new Record(i,T[i]));
        }
        return ansArr;
    }
}

File: leetcode/algorithms/stack/daily-temperatures.ts
/*
    leetcode
    category: "Algorithms"
    problem title: "Daily Temperatures"
*/

interface StackElement {
    value: number;
    index: number;
}

function dailyTemperatures(T: number[]): number[] {
    const stackArr: Array<StackElement> = [];
    const result: number[] = Array.from({
        length: T.length
    },() => 0);
    
    T.forEach((num,index) => {
        
        while(stackArr.length>0 && stackArr[stackArr.length-1].value<num){
            const element: StackElement = stackArr.pop();
            result[element.index] = index - element.index; 
        }
        
        stackArr.push({
            value: num,
            index: index
        })
        
    })
    
    return result;
    
    
};